### PR TITLE
Logging

### PR DESCRIPTION
**Pull request by Juxtaposedwords**
Moving to using the JSON logger. 

In the future, if/when we do move to events we can use this. For the mean time it provides an info log format for each service the logger runs on.

### CODE REVIEW

File: out/automation/util/log-listener.js [hunk 1]
```diff
@@ -1,33 +1,48 @@
...
+ *  async function log(ns, message) { await jsonLog(ns, "MY_COMPLETLEY_REAL_PROGRAM_NAME", message)} 
  * 
- * @param {import("../../..").NS } ns */
+ *  @param {import("../../..").NS } ns */
 export async function main(ns) {
-    ns.disableLog('ALL')
-    const h = ns.getPortHandle(1);
+    const data = ns.flags([
+		["port", 2], // Which port the listener should listen to.
+        ["log_destination", "home"], // Which port the listener should listen to.
```
Comment (uhhhclem) [inline out/automation/util/log-listener.js:17]: This comment's not right.

File: out/automation/util/log-listener.js [hunk 2]
```diff
@@ -1,33 +1,48 @@
...
+ *          await log(ns, "for the longest time.");
+ *      }
+ *  async function log(ns, message) { await jsonLog(ns, "MY_COMPLETLEY_REAL_PROGRAM_NAME", message)} 
  * 
- * @param {import("../../..").NS } ns */
+ *  @param {import("../../..").NS } ns */
 export async function main(ns) {
-    ns.disableLog('ALL')
-    const h = ns.getPortHandle(1);
+    const data = ns.flags([
```
Comment (uhhhclem) [inline out/automation/util/log-listener.js:15]: Let's call this const flags, not data.  Have you heard my rant about why no variable should ever be named data?

File: out/automation/util/log-listener.js [hunk 3]
```diff
@@ -1,33 +1,48 @@
...
+		["port", 2], // Which port the listener should listen to.
+        ["log_destination", "home"], // Which port the listener should listen to.
+	]);
+    const port = ns.getPortHandle(data["port"]);
```
Comment (uhhhclem) [inline out/automation/util/log-listener.js:19]: flags.port, not data["port"].  all keys in JS objects are accessible as properties.

File: out/automation/util/log-listener.js [hunk 4]
```diff
@@ -1,33 +1,48 @@
...
-            // let line = `${entry.date} ${entry.method}.js : target=${entry.target} start=${entry.} amount=${entry.amount}`
-            await ns.write(`logs/${entry.host}.log`, v + '\n')
-            await ns.write('logs/all.log', input + '\n')
-        } else {
-            // this is the current non-json behavior route.
-            await ns.write('log.txt', v + '\n');
-        }
+        let v = port.read();
+
+        let entry = JSON.parse(v);
```
Comment (uhhhclem) [inline out/automation/util/log-listener.js:28]: Both v and entry can be const, as they're not modified.

File: out/automation/util/log-listener.js [hunk 5]
```diff
@@ -1,33 +1,48 @@
...
+
+        // Write the raw json log.
+        const rawFile = `/logs/raw/${entry.host}.txt`;
+        // we can't use a variable for the "a" in this context, as the letter a is not the flag append.
+        update(ns, rawFile, v + "\n")
+
+        // Write out the structured log,
+        const outputFile = `/logs/${entry.host}/${entry.program}.txt`
+        let line = `${entry.datetime} ${entry.program} : ${entry.message}\n`
```
Comment (uhhhclem) [inline out/automation/util/log-listener.js:37]: Also can be const.

File: out/automation/util/log-listener.js [hunk 6]
```diff
@@ -1,33 +1,48 @@
...
-            // this is the current non-json behavior route.
-            await ns.write('log.txt', v + '\n');
-        }
+        let v = port.read();
+
+        let entry = JSON.parse(v);
+
+        // Write the raw json log.
+        const rawFile = `/logs/raw/${entry.host}.txt`;
+        // we can't use a variable for the "a" in this context, as the letter a is not the flag append.
```
Comment (uhhhclem) [inline out/automation/util/log-listener.js:32]: Is this comment outdated?

File: out/automation/util/log-listener.js [hunk 7]
```diff
@@ -1,33 +1,48 @@
...
+        // Write out the structured log,
+        const outputFile = `/logs/${entry.host}/${entry.program}.txt`
+        let line = `${entry.datetime} ${entry.program} : ${entry.message}\n`
+        update(ns, outputFile, line)
+    }
+}
+// TODO add log file truncation
+async function update(ns, fileName, line, destination) {
+    if (ns.fileExists(fileName)) {
+        await ns.write(fileName, line, "a")
```
Comment (uhhhclem) [inline out/automation/util/log-listener.js:44]: You'll hate this, but:

await ns.write(filename, line, ns.fileExists(filename) ? "a" : "w")

File: out/automation/util/weaken.js [hunk 8]
```diff
@@ -14,14 +14,18 @@ export async function main(ns) {
...
+	}
+    await log(`Started weaking: ${target}`);
  
     // Infinite loop that continously weakens the target server.
     while (true) {
         // Always sleep in an infinite loop.
         await ns.sleep(100);
         const amt = await ns.weaken(target)
-        await log(ns, 'weaken', target, amt)
+        await log(`Weakend for ${amt}`)
```
Comment (uhhhclem) [inline out/automation/util/weaken.js:28]: Weakened

File: out/automation/util/log-listener.js (hunk 1 above)
Comment (Juxtaposedwords) [reply out/automation/util/log-listener.js:17]: thanks, removing that lingering line

File: out/automation/util/log-listener.js (hunk 2 above)
Comment (Juxtaposedwords) [reply out/automation/util/log-listener.js:15]: Done

Comment (Juxtaposedwords) [reply out/automation/util/log-listener.js:15]: and no #suscribe

File: out/automation/util/log-listener.js (hunk 3 above)
Comment (Juxtaposedwords) [reply out/automation/util/log-listener.js:19]: done

File: out/automation/util/log-listener.js (hunk 4 above)
Comment (Juxtaposedwords) [reply out/automation/util/log-listener.js:28]: Done. but how will they ever grow as variables?

File: out/automation/util/log-listener.js (hunk 6 above)
Comment (Juxtaposedwords) [reply out/automation/util/log-listener.js:32]: yup, striking.

File: out/automation/util/log-listener.js (hunk 7 above)
Comment (Juxtaposedwords) [reply out/automation/util/log-listener.js:44]: Better solution for now. In the future i'd like to refactor so we truncate 1000 log files (like info logs does)

**Pull request merged**